discovery/gossiper.go

```diff
@@ -795,78 +813,28 @@ func (d *deDupedAnnouncements) Emit() []msgWithSenders {
 // associated channels still not have the full channel proofs assembled, we
 // will try to resend them. If we have the full proof, we can safely delete the
 // message from the messageStore.
-func (d *AuthenticatedGossiper) resendAnnounceSignatures() error {
-	peerMsgsToResend, err := d.cfg.MessageStore.Messages()
+func (d *AuthenticatedGossiper) resendPendingMsgs() error {
+	// Fetch all the messages that we should retry to send from the
+	// database.
+	pendingMsgs, err := d.cfg.MessageStore.Messages()
 	if err != nil {
 		return err
 	}
 
-	// We now iterate over these messages, resending those that we don't
-	// have the full proof for, deleting the rest.
-	for peer, msgsToResend := range peerMsgsToResend {
-		pubKey, err := btcec.ParsePubKey(peer[:], btcec.S256())
-		if err != nil {
-			return err
-		}
-
-		for _, msg := range msgsToResend {
-			msg := msg.(*lnwire.AnnounceSignatures)
-
-			// Check if the full channel proof exists in our graph.
-			chanInfo, _, _, err := d.cfg.Router.GetChannelByID(
-				msg.ShortChannelID)
-			if err != nil {
-				// If the channel cannot be found, it is most likely a
-				// leftover message for a channel that was closed.  In
-				// this case we delete it from the message store.
-				log.Warnf("unable to fetch channel info for "+
-					"chanID=%v from graph: %v. Will delete local"+
-					"proof from database",
-					msg.ChannelID, err)
-				err = d.cfg.MessageStore.DeleteMessage(msg, peer)
-				if err != nil {
-					return err
-				}
-				continue
-			}
-
-			// 1. If the full proof does not exist in the graph, it means
-			// that we haven't received the remote proof yet (or that we
-			// crashed before able to assemble the full proof). Since the
-			// remote node might think they have delivered their proof to
-			// us, we will resend _our_ proof to trigger a resend on their
-			// part: they will then be able to assemble and send us the
-			// full proof.
-			if chanInfo.AuthProof == nil {
-				err := d.sendAnnSigReliably(msg, pubKey)
-				if err != nil {
-					return err
-				}
-				continue
-			}
-
-			// 2. If the proof does exist in the graph, we have
-			// successfully received the remote proof and assembled the
-			// full proof. In this case we can safely delete the local
-			// proof from the database. In case the remote hasn't been able
-			// to assemble the full proof yet (maybe because of a crash),
-			// we will send them the full proof if we notice that they
-			// retry sending their half proof.
-			if chanInfo.AuthProof != nil {
-				log.Debugf("Deleting message for chanID=%v from "+
-					"messageStore", msg.ChannelID)
-				err := d.cfg.MessageStore.DeleteMessage(msg, peer)
-				if err != nil {
-					return err
-				}
+	// With the messages retrieved, we can now attempt to resend them.
+	for peerPubKey, peerMsgs := range pendingMsgs {
+		for _, msg := range peerMsgs {
+			if err := d.sendMsg(msg, peerPubKey); err != nil {
```

Comment:
What do you mean by unresponsive peer? The `SendMessage` call being used within `sendMsg` is sending the message asynchronously, so it won't wait for a response.

Reply:
Ok, so there's no way any of the `SendMessage` calls could block or error out? Feels a bit brittle, as any error returned (I'm thinking if an already connected peer goes offline, stops responding, or some other online<->offline transition) or unresponsiveness will halt gossiper startup.

Reply:
> so there's no way any of the SendMessage calls could block or error out?

Correct, there's no way it can block, but it can error out with `ErrPeerExiting`. See [here](https://github.com/lightningnetwork/lnd/blob/master/peer.go#L2211). Therefore, I think the goroutine isn't necessary here, unless we opt to block on the send.